core/src/main/scala/com/metamx/tranquility/tranquilizer/Tranquilizer.scala

```diff
@@ -0,0 +1,310 @@
+  maxBatchSize: Int,
+  maxPendingBatches: Int,
+  lingerMillis: Long
+) extends Logging
+{
+  require(maxBatchSize >= 0, "batchSize >= 0")
+  require(maxPendingBatches >= 1, "maxPendingBatches >= 1")
+  require(lingerMillis >= 0, "lingerMillis >= 0")
+
+  @volatile private var started: Boolean = false
+
+  case class MessageHolder(message: MessageType, future: Promise[Unit])
+
+  private val lock             : AnyRef                = new AnyRef
+  private var buffer           : Buffer[MessageHolder] = new ArrayBuffer[MessageHolder]
+  private var bufferStartMillis: Long                  = 0L
+  private var pendingBatches   : Int                   = 0
+
+  private val sendThread = loggingThread {
+    try {
+      while (!Thread.currentThread().isInterrupted) {
+        val exBuffer = lock.synchronized {
+          while (
+            buffer.isEmpty ||
+              pendingBatches >= maxPendingBatches ||
+              (lingerMillis > 0 && System.currentTimeMillis() < bufferStartMillis + lingerMillis)
+          ) {
+            if (lingerMillis == 0) {
+              lock.wait()
+            } else {
+              lock.wait(math.max(1, bufferStartMillis + lingerMillis - System.currentTimeMillis()))
+            }
+          }
+          if (log.isDebugEnabled) {
+            log.debug(s"Buffer with ${buffer.size} messages has lingered too long, sending.")
+          }
+
+          assert(buffer.nonEmpty)
+          assert(pendingBatches < maxPendingBatches)
+          swap()
+        }
+
+        sendBuffer(exBuffer)
+      }
+    }
+    catch {
+      case e: InterruptedException =>
+        log.debug("Interrupted, exiting thread.")
+    }
+  } withEffect { t =>
+    t.setDaemon(true)
+    t.setName(s"Tranquilizer-BackgroundSend-[$beam]")
+  }
+
+  /**
+    * Start the tranquilizer. Must be called before any calls to "send".
+    */
+  @LifecycleStart
+  def start(): Unit = {
+    started = true
+    sendThread.start()
+  }
+
+  /**
+    * Sends a message. Generally asynchronous, but will block if the internal buffer is full.
+    *
+    * The future may contain a Unit, in which case the message was successfully sent. Or it may contain an exception,
+    * in which case the message may or may not have been successfully sent. One specific exception to look out for is
+    * MessageDroppedException, which means that a message was dropped due to unrecoverable reasons. With Druid this
+    * can be caused by message timestamps being outside the configured windowPeriod.
+    *
+    * NB: This can actually be wrong: if *any* message in a batch of messages is dropped, then all messages in that
+    * batch will be marked as dropped (even if they actually made it out). This is a limitation of how Tranquility
+    * tracks which messages were and were not dropped. This could be improved once
+    * https://github.com/druid-io/tranquility/issues/56 is addressed.
+    */
+  def send(message: MessageType): Future[Unit] = {
+    requireStarted()
```

Comment:
was wondering about the behavior of the simple adapter with this as it will mess up with the counters that it maintains., 
I am fine with the exception itself, lets also mention it in javadoc.